nexus/src/app/background/sync_switch_port_settings.rs

```diff
@@ -0,0 +1,1236 @@
+                            continue;
+                        },
+                    };
+
+                    let request_prefixes = match bgp_announce_prefixes.get(&request_bgp_config.bgp_announce_set_id) {
+                        Some(prefixes) => prefixes,
+                        None => {
+                            error!(
+                                log,
+                                "no prefixes to announce found for bgp config";
+                                "switch" => ?location,
+                                "announce_set_id" => ?request_bgp_config.bgp_announce_set_id,
+                                "bgp_config_id" => ?config_id,
+                            );
+                            continue;
+                        },
+                    };
+
+                    let request = ApplyRequest {
+                        asn: *request_bgp_config.asn,
+                        peers,
+                        originate: request_prefixes.clone(),
+                    };
+
+                    match desired_bgp_configs.entry(*location) {
+                        Entry::Occupied(mut occupied_entry) => {
+                            occupied_entry.get_mut().push(request);
+                        }
+                        Entry::Vacant(vacant_entry) => {
+                            vacant_entry.insert(vec![request]);
+                        }
+                    }
+                }
+
+                for (location, configs) in &desired_bgp_configs {
+                    let client = match mgd_clients.get(location) {
+                        Some(client) => client,
+                        None => {
+                            error!(log, "no mgd client found for switch"; "switch_location" => ?location);
+                            continue;
+                        },
+                    };
+                    for config in configs {
+                        info!(
+                            &log,
+                            "applying bgp config";
+                            "switch_location" => ?location,
+                            "config" => ?config,
+                        );
+                        if let Err(e) = client.inner.bgp_apply(config).await {
+                            error!(log, "error while applying bgp configuration"; "error" => ?e);
+                        }
+                    }
+                }
+
+                //
+                // calculate and apply bootstore changes
+                //
+
+                // check downstream bootstore version
+                let mut downstream_config: Option<EarlyNetworkConfig> = None;
+
+                // Since we update the first scrimlet we can reach (we failover to the second one
+                // if updating the first one fails) we need to check them both.
+                for (_location, client) in &sled_agent_clients {
+                    let scrimlet_cfg  = match client.read_network_bootstore_config_cache().await {
```

Comment:
I don't think it's correct to read a bootstore config from a sled agent and then use the generation number as the basis for whether we need to bump the config as we're doing below with the `bootstore_needs_update` check. The sole source of truth is the nexus DB and the bootstore/sled agents below are in a constant state of eventual consistency convergence, so we should really be treating them unconditional write-only entities. @andrewjstone can keep me honest here.

Reply:
@andrewjstone and I actually discussed this a bit last week, but it's probably good to also discuss it here. The challenge we currently have is that we don't know _when_ the "db bootstore config" has actually drifted from what's on the bootstore, so we don't know when to bump the version number and push a change. Previously this logic was in a saga, and the saga only ran when there was a change, so we always bumped the number and pushed a new bootstore config to the sled agents. Here, we run every N seconds even if there wasn't a change. 

> I don't think it's correct to read a bootstore config from a sled agent and then use the generation number as the basis for whether we need to bump the config as we're doing below with the bootstore_needs_update check. 

So comparing the version is actually a mistake that explains a bug I just noticed! 😂 

The rest of the logic in that block was to address a shortcoming we have with our bootstore config tracking in Nexus:

The logic here was _intended_ to check both scrimlets to see who has the highest config, and then compare the other contents of what's in the db (ntp_servers & rack_network_config) against that config to see if a drift has occurred. IIUC, there _shouldn't_ be a scenario where the scrimlets are lagging behind any other sleds. Admittedly it's a bit of a hack. If we want to properly track changes to the bootstore config we'll probably need to add another table)

The bug I noticed is that the RPW is _always_ bumping the version number and updating the bootstore config, even if it hasn't changed, because our "db bootstore config" is being initialized with a version of `0` and then set to the next generation number via the `bump_version` datastore method if the RPW believes the bootstore needs to be updated (which it believes to be true because it's comparing the wrong field). This still results in a system that converges to the correct settings, but it means that the sled-agents a propagating bootstore configs far more often than they need to. 😂

Reply:
I forgot to follow up here. Levon and I did chat about this, and what he was doing reading from the bootstore was safe, but it's still not how it was intended to be used. Really, all reads and writes should go through CRDB to linearize them and then they should be pushed to bootstore. Levon is going to take a crack at adding a table for the bootstore data so a history of updates can be maintained and then only writing to bootstore.